src/Codegen/Constraints/UntypedBuilder.php

```diff
@@ -191,6 +191,158 @@ private function generateOneOfChecks(vec<TSchema> $schemas, HackBuilder $hb): vo
+   * bail out and just used `mixed`.
+   */
+  public function getMergedAllOfChecks(): ?TSchema {
+    // Bail if this schema contains more than an `allOf` constraint.
+    if (C\count(Shapes::toDict($this->typed_schema)) !== 1) {
+      return null;
+    }
+
+    $merged_properties = dict[];
+    $merged_required = vec[];
+    $max_properties = null;
+    $min_properties = null;
+    $coerce = false;
+
+    $schemas = $this->typed_schema['allOf'] ?? vec[]
+      |> Vec\reverse($$); // Reverse for parity with non-strict output; this shouldn't actually matter.
+    foreach ($schemas as $index => $schema) {
+      $schema_builder = new SchemaBuilder(
+        $this->ctx,
+        $this->generateClassName($this->suffix, 'allOf', (string)$index),
+        $schema
+      );
+      $schema = $schema_builder->getResolvedSchema();
+      if (Shapes::keyExists($schema, 'allOf')) {
+        $builder = new UntypedBuilder($this->ctx, 'allOf', $schema);
+        $schema = $builder->getMergedAllOfChecks();
+      }
+
+      if (!$schema || Shapes::idx($schema, 'type') !== TSchemaType::OBJECT_T) {
+        return null;
+      }
+
+      $schema = type_assert_shape<TObjectSchema>($schema, TObjectSchema::class);
+
+      if (Shapes::idx($schema, 'additionalProperties', true)) {
+        // If any schema allows arbitrary properties, just bail out to keep this simple.
+        return null;
+      }
+
+      if (Shapes::idx($schema, 'patternProperties')) {
+        // We don't currently support merging pattern properties, though we probably could.
+        return null;
+      }
+
+      $resolved_context = $schema_builder->getResolvedContext();
+      $resolved_root_directory = $resolved_context->getRefsRootDirectory();
+      if ($resolved_root_directory is null) {
+        $resolved_root_directory = __DIR__;
+      } else {
+        $resolved_root_directory = \realpath($resolved_root_directory);
+      }
+
+      foreach ($schema['properties'] ?? dict[] as $name => $prop) {
+        if (C\contains_key($merged_properties, $name)) {
+          // TODO: We could more intelligently handle duplicate keys.
+          return null;
+        }
+
+        if (Shapes::keyExists($prop, '$ref')) {
+          $root_directory = $this->ctx->getRefsRootDirectory();
+          if ($root_directory is null) {
+            $root_directory = __DIR__;
+          } else {
+            $root_directory = \realpath($root_directory);
+          }
+
+          if ($resolved_root_directory !== $root_directory) {
+            // Disgusting hack: let's just rewind the entire root directory
+            $prop_root = $resolved_root_directory.'/';
+            if (Str\starts_with($prop['$ref'], '#')) {
+              // Ref in the same file
+              $prop_root = $resolved_root_directory.'/'.$resolved_context->getCurrentRefFileName();
+            } else {
+              // External file containing ref
+              $prop_root = $resolved_root_directory.'/';
```

Comment:
Do you need this else statement since you initialize `prop_root` with this?

Reply:
Good catch, removed